File: examples/starters/mastra/starter.ts
/**
 * ACP Starter — Mastra (TypeScript)
 *
 * The minimum code to wire ACP governance into a Mastra agent. Copy this
 * folder, swap the placeholder tool for your real one, ship.
 *
 * Governance pattern: decorator. `governed("tool_name", fn)` from the
 * base `@agenticcontrolplane/governance` package wraps the tool's
 * `execute` function with ACP's pre/post hooks. Mastra calls the wrapped
 * version; denials return a tool_error string that the agent sees as
 * tool output and adapts to.
 *
 * No framework-specific ACP adapter package is needed — base
 * `governed()` composes cleanly with Mastra's `createTool()`.
 *
 * Run:  bash run.sh
 */

import dotenv from "dotenv";
import path from "node:path";
import { fileURLToPath } from "node:url";

import { Mastra } from "@mastra/core";
import { Agent } from "@mastra/core/agent";
import { createTool } from "@mastra/core/tools";
import { z } from "zod";

import {
  configure,
  governed,
  withContext,
} from "@agenticcontrolplane/governance";

// ── 1. Load ACP credentials from ./.env next to this file.
const __dirname = path.dirname(fileURLToPath(import.meta.url));
dotenv.config({ path: path.join(__dirname, ".env") });

const { OPENAI_API_KEY, ACP_USER_TOKEN, ACP_GATEWAY_URL = "https://api.agenticcontrolplane.com" } = process.env;
if (!OPENAI_API_KEY) throw new Error("Missing OPENAI_API_KEY");
if (!ACP_USER_TOKEN) throw new Error("Missing ACP_USER_TOKEN (gsk_...)");

// ── 2. Point the governance SDK at your ACP gateway. Once per process.
configure({ baseUrl: ACP_GATEWAY_URL });

// ── 3. Your tool. `governed(name, fn)` wraps the execute function with
// ACP's pre/post hook protocol. Mastra's `createTool({ execute })` is
// passed the wrapped function — governance is invisible to Mastra.
// REPLACE the body with your real logic.
const lookupRecord = createTool({
  id: "lookup_record",
  description: "Look up a record by ID. Replace with your real tool description.",
  inputSchema: z.object({
    id: z.string().describe("The record ID to look up."),
  }),
  outputSchema: z.object({
    id: z.string(),
    status: z.string(),
    note: z.string(),
  }),
  execute: governed("lookup_record", async ({ context }) => {
    return { id: context.id, status: "placeholder", note: "replace me with real logic" };
  }),
});

// ── 4. Define the agent. Mastra's model router parses the prefix —
// "openai/..." uses OPENAI_API_KEY; swap to "anthropic/claude-sonnet-4-6"
// + ANTHROPIC_API_KEY if you'd rather.
const agent = new Agent({
  id: "my-mastra-agent",
  name: "My Mastra Agent",
  instructions: "You are an ACP-governed agent. Use the tools available.",
  model: "openai/gpt-4o-mini",
  tools: { lookupRecord },
});

const mastra = new Mastra({ agents: { agent } });

// ── 5. `withContext` binds user identity for every governance call
// inside. Without it, pre/post hooks silently no-op.
async function main() {
  await withContext(
    {
      userToken: ACP_USER_TOKEN!,
      agentName: "my-mastra-agent",
      agentTier: "background",
    },
    async () => {
      const res = await mastra.getAgentById("my-mastra-agent")!.generate(
        "Look up record id=abc-123 and tell me what you find."
      );
      console.log(res.text);
    },
  );
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: examples/starters/pydantic-ai/starter.py
"""
ACP Starter — Pydantic AI (Python)

The minimum code to wire ACP governance into a Pydantic AI agent. Copy
this folder, swap the placeholder tool for your real one, ship.

Governance pattern: decorator. `@governed("tool_name")` is stacked under
Pydantic AI's `@agent.tool_plain`, so every tool call flows through ACP's
pre/post hooks. Denials return a tool_error string that Pydantic AI
delivers to the agent as tool output; the agent adapts.

Run:  bash run.sh
"""

from __future__ import annotations

import json
import os
from pathlib import Path

from dotenv import load_dotenv

from acp_governance import configure, governed, set_context
from pydantic_ai import Agent

# ── 1. Load ACP credentials from ./.env next to this file.
load_dotenv(Path(__file__).parent / ".env")
for v in ("ANTHROPIC_API_KEY", "ACP_USER_TOKEN"):
    if not os.environ.get(v):
        raise SystemExit(f"Missing {v} — copy .env.example → .env")

# ── 2. Point the governance SDK at your ACP gateway. Once per process.
configure(
    base_url=os.environ.get("ACP_GATEWAY_URL", "https://api.agenticcontrolplane.com"),
    client_header="acp-pydantic-ai-starter/0.1.0",
)

# ── 3. Build the agent. `Agent(model_string, ...)` resolves the provider
# from the prefix; ANTHROPIC_API_KEY is picked up by the Anthropic model.
# Swap `"anthropic:claude-sonnet-4-6"` for any provider-prefixed string
# (e.g. "openai:gpt-4o-mini") — see Pydantic AI docs.
agent = Agent(
    "anthropic:claude-sonnet-4-6",
    instructions="You are an ACP-governed agent. Use the tools available.",
)


# ── 4. Your tool. `@agent.tool_plain` registers it with Pydantic AI;
# `@governed` wraps the call in ACP's pre/post hooks. Order matters —
# @governed must be INSIDE @agent.tool_plain so Pydantic AI's signature
# introspection and the tool registration both see the wrapped function.
# REPLACE the body with your real logic (DB lookup, API call, etc.).
@agent.tool_plain
@governed("lookup_record")
def lookup_record(id: str) -> str:
    """Look up a record by ID. Replace with your real tool description."""
    return json.dumps({"id": id, "status": "placeholder", "note": "replace me"})


def main() -> None:
    # ── 5. Bind identity for every governance call inside. Without this
    # set_context call, ACP pre/post hooks silently no-op.
    set_context(
        user_token=os.environ["ACP_USER_TOKEN"],
        agent_name="my-pydantic-agent",  # Rename for your agent.
        agent_tier="background",
    )

    result = agent.run_sync("Look up record id=abc-123 and tell me what you find.")
    print(result.output)


if __name__ == "__main__":
    main()
